{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load modules\n",
    "from Bio import Entrez, Seq, SeqIO, pairwise2, Align\n",
    "import numpy as np\n",
    "\n",
    "#load nucleotide sequences from NCBI \n",
    "def get_seq(db, key, rettype=\"gb\"):\n",
    "    Entrez.email = \"A.N.Other@example.com\"\n",
    "    handle = Entrez.efetch(db=db, id=key, rettype=rettype)\n",
    "    return SeqIO.read(handle, \"gb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Call NCBI API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#[welche daten werden geladen?]\n",
    "#hvr-1\n",
    "nea_1a = get_seq(\"nucleotide\", \"AF011222\")\n",
    "nea_1b = get_seq(\"nucleotide\", \"AF282971\")\n",
    "\n",
    "#[welche daten werden geladen?]\n",
    "#hvr-2\n",
    "nea_2a = get_seq(\"nucleotide\", \"AF142095\")\n",
    "nea_2b = get_seq(\"nucleotide\", \"AF282972\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 206 #known length\n",
    "\n",
    "nea_1 = nea_1a.seq + nea_1b.seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local Alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "664.0\n",
      "665.0\n",
      "665.0\n",
      "666.0\n",
      "664.0\n",
      "668.0\n",
      "663.0\n",
      "663.0\n",
      "662.0\n",
      "662.0\n",
      "662.0\n",
      "662.0\n",
      "662.0\n",
      "662.0\n",
      "664.0\n",
      "665.0\n",
      "666.0\n",
      "665.0\n",
      "667.0\n",
      "669.0\n",
      "670.0\n",
      "665.0\n",
      "662.0\n",
      "666.0\n",
      "664.0\n",
      "665.0\n",
      "661.0\n",
      "668.0\n",
      "666.0\n",
      "666.0\n",
      "664.0\n",
      "663.0\n",
      "667.0\n",
      "664.0\n",
      "662.0\n",
      "667.0\n",
      "662.0\n",
      "663.0\n",
      "661.0\n",
      "663.0\n",
      "664.0\n",
      "663.0\n",
      "663.0\n",
      "664.0\n",
      "667.0\n",
      "665.0\n",
      "665.0\n",
      "662.0\n",
      "662.0\n",
      "661.0\n",
      "665.0\n",
      "667.0\n",
      "670.0\n",
      "665.0\n",
      "666.0\n",
      "667.0\n",
      "666.0\n",
      "666.0\n",
      "662.0\n",
      "665.0\n",
      "663.0\n",
      "665.0\n",
      "665.0\n",
      "667.0\n",
      "662.0\n",
      "662.0\n",
      "666.0\n",
      "664.0\n",
      "666.0\n",
      "666.0\n",
      "661.0\n",
      "663.0\n",
      "666.0\n",
      "665.0\n",
      "665.0\n",
      "664.0\n",
      "662.0\n",
      "665.0\n",
      "664.0\n",
      "665.0\n",
      "666.0\n",
      "663.0\n",
      "666.0\n",
      "666.0\n",
      "665.0\n",
      "667.0\n",
      "664.0\n",
      "664.0\n",
      "667.0\n",
      "667.0\n",
      "665.0\n",
      "663.0\n",
      "665.0\n",
      "667.0\n",
      "660.0\n",
      "665.0\n",
      "665.0\n",
      "669.0\n",
      "664.0\n",
      "666.0\n",
      "667.0\n",
      "664.0\n",
      "669.0\n",
      "666.0\n",
      "664.0\n",
      "664.0\n",
      "665.0\n",
      "665.0\n",
      "665.0\n",
      "662.0\n",
      "666.0\n",
      "664.0\n",
      "664.0\n",
      "667.0\n",
      "666.0\n",
      "666.0\n",
      "667.0\n",
      "664.0\n",
      "664.0\n",
      "664.0\n",
      "666.0\n",
      "666.0\n",
      "665.0\n",
      "663.0\n",
      "664.0\n",
      "666.0\n",
      "666.0\n",
      "663.0\n",
      "663.0\n",
      "665.0\n",
      "661.0\n",
      "667.0\n",
      "665.0\n",
      "662.0\n",
      "668.0\n",
      "667.0\n",
      "663.0\n",
      "667.0\n",
      "664.0\n",
      "664.0\n",
      "668.0\n",
      "666.0\n",
      "669.0\n",
      "661.0\n",
      "667.0\n",
      "668.0\n",
      "661.0\n",
      "661.0\n",
      "668.0\n",
      "665.0\n",
      "661.0\n",
      "668.0\n",
      "667.0\n",
      "669.0\n",
      "666.0\n",
      "664.0\n",
      "668.0\n",
      "670.0\n",
      "662.0\n",
      "665.0\n",
      "665.0\n",
      "666.0\n",
      "667.0\n",
      "663.0\n",
      "670.0\n",
      "668.0\n",
      "668.0\n",
      "665.0\n",
      "665.0\n",
      "668.0\n",
      "665.0\n",
      "665.0\n",
      "661.0\n",
      "665.0\n",
      "666.0\n",
      "663.0\n",
      "663.0\n",
      "664.0\n",
      "664.0\n",
      "664.0\n",
      "666.0\n",
      "668.0\n",
      "666.0\n",
      "665.0\n",
      "666.0\n",
      "666.0\n",
      "667.0\n",
      "670.0\n",
      "667.0\n",
      "665.0\n",
      "662.0\n",
      "668.0\n",
      "670.0\n",
      "667.0\n",
      "667.0\n",
      "665.0\n",
      "664.0\n",
      "661.0\n",
      "667.0\n",
      "664.0\n",
      "661.0\n",
      "668.0\n",
      "668.0\n",
      "667.0\n",
      "665.0\n",
      "664.0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\nfrom Bio import AlignIO\\nhandle = ...\\nfor a in AlignIO.parse(handle, \"fasta-m10\"):\\n  assert len(a) == 2, \"Should be pairwise!\"\\n  print(\"Alignment length %i\" % a.get_alignment_length())\\n  for record in a:\\n    print(\"%s %s %s\" % (record.seq, record.name, record.id))\\n-----------------------------------------------------------------\\n>>> from Bio import Align\\n>>> aligner = Align.PairwiseAligner()\\n>>> alignments = aligner.align(\"ACCGT\", \"ACG\")\\n>>> for alignment in sorted(alignments):\\n...     print(\"Score = %.1f:\" % alignment.score)\\n...     print(alignment)'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score = np.zeros(n)\n",
    "startat = np.zeros(n)\n",
    "\n",
    "#for i=1:n\n",
    "#      [sc,alignment,st]=swalign(human_dloop(i).Sequence,nea.Sequence,'Alphabet', 'NT');\n",
    "aligner = Align.PairwiseAligner()\n",
    "aligner.mode = 'local'\n",
    "with open(\"d_loop.fasta\", \"r\") as handle:\n",
    "    alignments = []\n",
    "    d_loop = SeqIO.parse(handle, \"fasta\")\n",
    "    for seq in d_loop:\n",
    "        a = aligner.align(nea_1, seq.seq)\n",
    "        print(a.score)\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "            \n",
    "            \n",
    "\"\"\"\n",
    "from Bio import AlignIO\n",
    "handle = ...\n",
    "for a in AlignIO.parse(handle, \"fasta-m10\"):\n",
    "  assert len(a) == 2, \"Should be pairwise!\"\n",
    "  print(\"Alignment length %i\" % a.get_alignment_length())\n",
    "  for record in a:\n",
    "    print(\"%s %s %s\" % (record.seq, record.name, record.id))\n",
    "-----------------------------------------------------------------\n",
    ">>> from Bio import Align\n",
    ">>> aligner = Align.PairwiseAligner()\n",
    ">>> alignments = aligner.align(\"ACCGT\", \"ACG\")\n",
    ">>> for alignment in sorted(alignments):\n",
    "...     print(\"Score = %.1f:\" % alignment.score)\n",
    "...     print(alignment)\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GTTCTTTCATGGGGGAGCAGATTTGGGTACCACCCAAGTATTGACTCACCCATC...CCC', IUPACAmbiguousDNA())"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nea_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[664.0, 665.0, 665.0, 666.0, 664.0, 668.0, 663.0, 663.0, 662.0, 662.0, 662.0, 662.0, 662.0, 662.0, 664.0, 665.0, 666.0, 665.0, 667.0, 669.0, 670.0, 665.0, 662.0, 666.0, 664.0, 665.0, 661.0, 668.0, 666.0, 666.0, 664.0, 663.0, 667.0, 664.0, 662.0, 667.0, 662.0, 663.0, 661.0, 663.0, 664.0, 663.0, 663.0, 664.0, 667.0, 665.0, 665.0, 662.0, 662.0, 661.0, 665.0, 667.0, 670.0, 665.0, 666.0, 667.0, 666.0, 666.0, 662.0, 665.0, 663.0, 665.0, 665.0, 667.0, 662.0, 662.0, 666.0, 664.0, 666.0, 666.0, 661.0, 663.0, 666.0, 665.0, 665.0, 664.0, 662.0, 665.0, 664.0, 665.0, 666.0, 663.0, 666.0, 666.0, 665.0, 667.0, 664.0, 664.0, 667.0, 667.0, 665.0, 663.0, 665.0, 667.0, 660.0, 665.0, 665.0, 669.0, 664.0, 666.0, 667.0, 664.0, 669.0, 666.0, 664.0, 664.0, 665.0, 665.0, 665.0, 662.0, 666.0, 664.0, 664.0, 667.0, 666.0, 666.0, 667.0, 664.0, 664.0, 664.0, 666.0, 666.0, 665.0, 663.0, 664.0, 666.0, 666.0, 663.0, 663.0, 665.0, 661.0, 667.0, 665.0, 662.0, 668.0, 667.0, 663.0, 667.0, 664.0, 664.0, 668.0, 666.0, 669.0, 661.0, 667.0, 668.0, 661.0, 661.0, 668.0, 665.0, 661.0, 668.0, 667.0, 669.0, 666.0, 664.0, 668.0, 670.0, 662.0, 665.0, 665.0, 666.0, 667.0, 663.0, 670.0, 668.0, 668.0, 665.0, 665.0, 668.0, 665.0, 665.0, 661.0, 665.0, 666.0, 663.0, 663.0, 664.0, 664.0, 664.0, 666.0, 668.0, 666.0, 665.0, 666.0, 666.0, 667.0, 670.0, 667.0, 665.0, 662.0, 668.0, 670.0, 667.0, 667.0, 665.0, 664.0, 661.0, 667.0, 664.0, 661.0, 668.0, 668.0, 667.0, 665.0, 664.0]\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
